File: python/sort/mergesort.py
def MERGE(a, b):
	c = []
	i, j = 0, 0
	while i < len(a) and j < len(b):
		if a[i] < b[j]:
			c.append(a[i])
			i += 1
		else:
			c.append(b[j])
			j += 1
	while i < len(a):
		c.append(a[i])
		i += 1
	while j < len(b):
		c.append(b[j])
		j += 1
	return c


def MERGE_SORT(A):
	n = len(A)
	if n < 2:
		a = A[0:n/2]
		b = A[n/2:]
	else:
		a = MERGE_SORT(A[0:n/2])
		b = MERGE_SORT(A[n/2:])
	return MERGE(a, b)

File: java/sort/MergeSort.java
class MergeSort
{
  private static int helper[];
  public static void mergeSort(int a[])
  {
    helper=new int[a.length];
    msort(a,0,a.length-1);
  }
  private static void msort(int a[], int l, int r)
  {
    if(l<r)
    {
      msort(a,l,(r+l)/2);
      msort(a,((r+l)/2)+1,r);
      merge(a,l,r);
    }
  }
  private static void merge(int a[], int l, int r)
  {
    int i,lc=l,rc=((r+l)/2)+1 ,hc=l;

    for(i=l;i<=r;i++)
    {
      helper[i]=a[i];
    }
    while(lc<=(r+l)/2 && rc<=r)
    {
      if(helper[lc]<=helper[rc])
      {
        a[hc++]=helper[lc++];
      }
      else
      {
        a[hc++]=helper[rc++];
      }
    }
    while(lc<=(r+l)/2)
    {
      a[hc++]=helper[lc++];
    }
  }
}
